### ISSUE TITLE
Eclipse-Resolver: Comparison method violates its general contract

### ISSUE THREAD

**Issue by mwgmnn** [eclipse-equinox/equinox#156]
We use the "old" but fast Eclipse resolver in one of our projects and got this exception lately. It is triggered by the commons-io - bundle which has the specialty to export its packages twice. This is ok, as is described here for example: https://bnd.discourse.group/t/why-is-commons-io-2-6-exports-2-versions-of-same-package-1-4-and-2-6/94

```
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
    at java.base/java.util.TimSort.mergeHi(TimSort.java:903)
    at java.base/java.util.TimSort.mergeAt(TimSort.java:520)
    at java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
    at java.base/java.util.TimSort.sort(TimSort.java:245)
    at java.base/java.util.Arrays.sort(Arrays.java:1515)
    at java.base/java.util.ArrayList.sort(ArrayList.java:1750)
    at java.base/java.util.Collections.sort(Collections.java:179)
    at org.eclipse.osgi.internal.module.VersionHashMap.reorder(VersionHashMap.java:87)
    at org.eclipse.osgi.internal.module.ResolverImpl.reorderGenerics(ResolverImpl.java:2356)
    at org.eclipse.osgi.internal.module.ResolverImpl.resolve(ResolverImpl.java:539)
    at org.eclipse.osgi.internal.resolver.StateImpl.resolve(StateImpl.java:601)
    at org.eclipse.osgi.internal.resolver.StateImpl.resolve(StateImpl.java:661)
```

Maybe we replace the last code line in VersionHashMap.compare by this one?:
`return Long.compare(vs1.getBundleDescription().getBundleId(), vs2.getBundleDescription().getBundleId());`

```
	// Compares two VersionSuppliers for descending ordered sorts.
	// The VersionSuppliers are sorted by the following priorities
	// First the resolution status of the supplying bundle.
	// Second is the supplier version.
	// Third is the bundle id of the supplying bundle.
	@Override
	public int compare(V vs1, V vs2) {
		// if the selection policy is set then use that
		if (resolver.getSelectionPolicy() != null)
			return resolver.getSelectionPolicy().compare(vs1.getBaseDescription(), vs2.getBaseDescription());
		if (preferSystemPackages) {
			String systemBundle = resolver.getSystemBundle();
			if (systemBundle.equals(vs1.getBundleDescription().getSymbolicName()) && !systemBundle.equals(vs2.getBundleDescription().getSymbolicName()))
				return -1;
			else if (!systemBundle.equals(vs1.getBundleDescription().getSymbolicName()) && systemBundle.equals(vs2.getBundleDescription().getSymbolicName()))
				return 1;
		}
		if (vs1.getBundleDescription().isResolved() != vs2.getBundleDescription().isResolved())
			return vs1.getBundleDescription().isResolved() ? -1 : 1;
		int versionCompare = -(vs1.getVersion().compareTo(vs2.getVersion()));
		if (versionCompare != 0)
			return versionCompare;
		return vs1.getBundleDescription().getBundleId() <= vs2.getBundleDescription().getBundleId() ? -1 : 1;
	}
```

**Comment by mickaelistria** [comment]
From what I understand, I believe this resolver is mostly unmaintained in Equinox and more a candidate for deletion or migration into PDE (which seems to be the only remaining thing using it in the Eclipse SDK project). So don't expect much fixes there.
@tjwatson do you confirm?

**Comment by tjwatson** [comment]
> So don't expect much fixes there. @tjwatson do you confirm?

Pretty accurate, that being said contributions will be considered.  I guess this scenario involves a bundle exporting the package multiple times but with the same version.  If you provide a testcase similar to the following that shows the problem and a contribution that fixes it I can review:

https://github.com/eclipse-equinox/equinox/blob/d1c1ac4ceab2f1ee4634335e5c1da455d877d7da/bundles/org.eclipse.osgi.tests/src/org/eclipse/osgi/tests/services/resolver/StateResolverTest.java#L4729-L4771

**Comment by mwgmnn** [comment]
It's not easy to reproduce because i don't know enough about the internal TimSort logic to find a simple test case. So I had to take and trim my real world example. The test triggers the exception not because of the Export-Package specialty that I mentioned, but because of the given Provide-Capability/osgi.service entries. I think this doesn't matter. And I also want to mention that this test case also might fail in the future if the sort algorithm in the JRE changes. Maybe it's better to leave it out when you consider to merge this pull request.

So thanks for your efforts. I understand that this resolver is deprecated but I tried to use the resolver service from Apache Felix and this was way too slow. I need the same user experience as with the resolution in the Eclipse-IDE. I dont't know the differences in the algorithms and for now i haven't seen any resolution problems.
